File: leetcode/Matrix/0037_Sudoku_Solver.cpp
class Solution {
private:
	vector<vector<int>> solve(vector<vector<int>> board) {
		for (int i = 0; i < 9; i++) {
			for (int j = 0; j < 9; j++) {
				if (board[i][j] == -1) {
					set<int> candidates({1, 2, 3, 4, 5, 6, 7, 8, 9});

					// check row and column
					for (int k = 0; k < 9; k++) {
						if (board[i][k] != -1 && candidates.find(board[i][k]) != candidates.end())
							candidates.erase(candidates.find(board[i][k]));
						if (board[k][j] != -1 && candidates.find(board[k][j]) != candidates.end())
							candidates.erase(candidates.find(board[k][j]));
					}
					// check sub-box
					for (int k = 0; k < 3; k++) {
						for (int l = 0; l < 3; l++) {
							int temp = board[(i / 3) * 3 + k][(j / 3) * 3 + l];
							if (temp != -1 && candidates.find(temp) != candidates.end())
								candidates.erase(candidates.find(temp));
						}
					}

					// try all candidates
					for (set<int>::iterator it = candidates.begin(); it != candidates.end(); it++) {
						board[i][j] = *it;
						vector<vector<int>> result = solve(board);

						// find if result has empty cell
						bool completed = true;
						for (int k = 0; k < 9; k++) {
							if (find(result[k].begin(), result[k].end(), -1) != result[k].end()) {
								completed = false;
								break;
							}
						}

						if (completed)
							return result;
					}
					return board;
				}
			}
		}
		return board;
	}
public:
    void solveSudoku(vector<vector<char>>& board) {
        vector<vector<int>> numBoard(9, vector<int>(9, -1));
        for (int i = 0; i < 9; i++) {
        	for (int j = 0; j < 9; j++) {
        		if (board[i][j] != '.')
        			numBoard[i][j] = board[i][j] - '0';
        	}
        }

        vector<vector<int>> result = solve(numBoard);
        for (int i = 0; i < 9; i++) {
        	for (int j = 0; j < 9; j++) {
        		board[i][j] = result[i][j] + '0';
        	}
        }
    }
};

File: leetcode/Matrix/0037_Sudoku_Solver.py
class Solution(object):
    def solve(self, board):
        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    candidate = set(['1', '2', '3', '4', '5', '6', '7', '8', '9'])
                    # check same row and column
                    for k in range(9):
                        if board[i][k] != '.' and board[i][k] in candidate:
                            candidate.remove(board[i][k])
                        if board[k][j] != '.' and board[k][j] in candidate:
                            candidate.remove(board[k][j])
                    # check sub-box
                    for k in range(3):
                        for l in range(3):
                            i2 = i // 3 * 3 + k
                            j2 = j // 3 * 3 + l
                            if board[i2][j2] != '.' and board[i2][j2] in candidate:
                                candidate.remove(board[i2][j2])

                    # try all candidate
                    if not candidate:
                        return None

                    for num in candidate:
                        board[i][j] = num
                        result = self.solve(board)
                        if result:
                            return result

                    # pass by reference, undo before return
                    board[i][j] = '.'
                    return None

        return board

    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        self.solve(board)
